### PR TITLE
BAP-506 - Support custom build name and number

### PR DESCRIPTION
**Pull request by barbelity**

### CODE REVIEW

File: src/main/java/org/jfrog/bamboo/builder/BuildInfoHelper.java [hunk 1]
```diff
@@ -196,10 +200,9 @@ private void addBuildParentProperties(Map<String, String> propertyMap, TriggerRe
         }
     }
 
-    private static Module createModule(TaskContext taskContext, List<Artifact> artifacts, List<Dependency> dependencies, Map<String, String> properties) throws IOException {
+    private static Module createModule(List<Artifact> artifacts, List<Dependency> dependencies, String buildName, String buildNumber) {
```
Comment (eyalbe4) [inline src/main/java/org/jfrog/bamboo/builder/BuildInfoHelper.java:203]: It makes sense for buildName and buildNumber to be the first args for this method.
The same goes for other methods in this class as well.

**Review by eyalbe4 — CHANGES_REQUESTED** [review]
I haven't finished reviewing this yet, but releasing what I have so far.

File: src/main/java/org/jfrog/bamboo/task/ArtifactoryPublishBuildInfoTask.java [hunk 2]
```diff
@@ -54,38 +62,101 @@ protected void initTask(@NotNull TaskContext context) {
...
+            Build build = buildInfoHelper.getBuilder(taskContext).build();
+            // Aggregate relevant builds to one build.
+            if (StringUtils.isNotBlank(aggregatedBuildsJson)) {
+                addBuildsToPublish(taskContext, build, aggregatedBuildsJson);
             }
             client.sendBuildInfo(build);
-            taskContext.getBuildContext().getBuildResult().getCustomBuildData().put(BUILD_RESULT_SELECTED_SERVER_PARAM, client.getArtifactoryUrl());
+
+            // Add build details to context.
+            addPublishedBuildDetailsToContext(client.getArtifactoryUrl(), build.getName(), build.getNumber(), taskContext, customBuildData);
```
Comment (eyalbe4) [inline src/main/java/org/jfrog/bamboo/task/ArtifactoryPublishBuildInfoTask.java:84]: By reading the above code, it is hard to understand the function of the ```addPublishedBuildDetailsToContet``` method vs. the above ```addBuildsToPublish``` method. Let's add further descriptions to make it easier to understand the code flow.

File: src/main/java/org/jfrog/bamboo/util/PublishedBuildDetails.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        this.buildUrl = buildUrl;
+    }
+
+    //    public String getBuildInfoUrl() {
+//        return this.artifactoryUrl + "/webapp/builds/" + this.buildName + "/" + this.buildNumber;
+//    }
+//
+//    public String getDisplayName() {
+//        return this.buildName + " / " + this.buildNumber;
+//    }
```
Comment (eyalbe4) [inline src/main/java/org/jfrog/bamboo/util/PublishedBuildDetails.java:62]: Let's remove these comments.

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

**Review by eyalbe4 — CHANGES_REQUESTED** [review]
I'm done reviewing everything. I'd love to review this again following the upcoming changes to this PR.

File: src/main/java/org/jfrog/bamboo/builder/BuildInfoHelper.java (hunk 1 above)
Comment (barbelity) [reply src/main/java/org/jfrog/bamboo/builder/BuildInfoHelper.java:203]: Fixed.

File: src/main/java/org/jfrog/bamboo/util/TaskUtils.java [hunk 4]
```diff
@@ -243,4 +245,66 @@ public static String getExecutablePath(AbstractBuildContext buildContext, Capabi
         }
         return env;
     }
+
```
[the remark this answers is not in the document]
Comment (barbelity) [reply src/main/java/org/jfrog/bamboo/util/TaskUtils.java:248]: Created a base class for all context classes - 'ArtifactoryBuildContext' with the required methods.
Removed the new methods from this class.

File: src/main/java/org/jfrog/bamboo/configuration/util/TaskConfigurationValidations.java [hunk 5]
```diff
@@ -143,4 +143,21 @@ public static void validateReleaseManagement(@NotNull ActionParametersMap params
             }
         }
     }
+
+    public static void validateCaptureBuildInfoParams(String buildNameKey, String buildNumberKey, String captureBuildInfoKey, @NotNull ActionParametersMap params, @NotNull ErrorCollection errorCollection) {
+        if (!params.getBoolean(captureBuildInfoKey)) {
+            return;
+        }
+        validateBuildNameNumber(buildNameKey, buildNumberKey, params, errorCollection);
```
[the remark this answers is not in the document]
Comment (barbelity) [reply src/main/java/org/jfrog/bamboo/configuration/util/TaskConfigurationValidations.java:151]: Changed.

File: src/main/java/org/jfrog/bamboo/util/PublishedBuildDetails.java (hunk 3 above)
Comment (barbelity) [reply src/main/java/org/jfrog/bamboo/util/PublishedBuildDetails.java:62]: Removed.

File: src/main/java/org/jfrog/bamboo/task/ArtifactoryPublishBuildInfoTask.java (hunk 2 above)
Comment (barbelity) [reply src/main/java/org/jfrog/bamboo/task/ArtifactoryPublishBuildInfoTask.java:84]: Please see my changes in the descriptions of these methods.

File: src/main/java/org/jfrog/bamboo/configuration/AbstractArtifactoryConfiguration.java [hunk 6]
```diff
@@ -345,4 +347,23 @@ private String getAuthTypeName(final String authType) {
     public void setI18nResolver(final I18nResolver i18nResolver) {
         this.i18nResolver = i18nResolver;
     }
+
+    public static void populateDefaultValuesInBuildContext(@NotNull Map<String, Object> context) {
+        String envVarsExcludePatterns = (String) context.get(AbstractBuildContext.ENV_VARS_EXCLUDE_PATTERNS);
```
[the remark this answers is not in the document]
Comment (barbelity) [reply src/main/java/org/jfrog/bamboo/configuration/AbstractArtifactoryConfiguration.java:352]: I wanted to make a method for populating the default values we have for our tasks.
Moved the population of envVarsExcludePatterns to a method of its own.

File: src/main/java/org/jfrog/bamboo/configuration/ArtifactoryCollectBuildIssuesConfiguration.java [hunk 7]
```diff
@@ -74,5 +77,8 @@ public boolean taskProducesTestResults(@NotNull TaskDefinition taskDefinition) {
     public void validate(@NotNull ActionParametersMap params, @NotNull ErrorCollection errorCollection) {
         // Validate publish server.
         TaskConfigurationValidations.validateArtifactoryServerProvidedAndValid(CollectBuildIssuesContext.SERVER_ID_PARAM, serverConfigManager, params, errorCollection);
+
+        // Validate build name and number.
+        TaskConfigurationValidations.validateBuildNameNumber(AbstractBuildContext.BUILD_NAME, AbstractBuildContext.BUILD_NUMBER, params, errorCollection);
```
[the remark this answers is not in the document]
Comment (barbelity) [reply src/main/java/org/jfrog/bamboo/configuration/ArtifactoryCollectBuildIssuesConfiguration.java:82]: Yes it makes the fields mandatory, as we have to know the exact build to attach the collected issues to.
The same goes with all other usages.
As not all classes which are extending the superclass require this validation, I didn't add it there.

File: src/main/java/org/jfrog/bamboo/configuration/AbstractGenericBuildConfiguration.java [hunk 8]
```diff
@@ -36,11 +40,19 @@ public void populateContextForEdit(@NotNull Map<String, Object> context, @NotNul
...
-        context.put("serverConfigManager", serverConfigManager);
-        context.put("selectedServerId", context.get(GenericContext.PREFIX + GenericContext.SERVER_ID_PARAM));
+        String buildName = (String)context.get(GenericContext.BUILD_NAME);
+        if (buildName == null) {
+            context.put(GenericContext.BUILD_NAME, AbstractBuildContext.DEFAULT_BUILD_NAME);
+        }
+        String buildNumber = (String)context.get(GenericContext.BUILD_NUMBER);
+        if (buildNumber == null) {
+            context.put(GenericContext.BUILD_NUMBER, AbstractBuildContext.DEFAULT_BUILD_NUMBER);
+        }
```
[the remark this answers is not in the document]
Comment (barbelity) [reply src/main/java/org/jfrog/bamboo/configuration/AbstractGenericBuildConfiguration.java:55]: Changed to use the super class' method 'populateDefaultBuildNameNumberInBuildContext()'.

File: src/main/java/org/jfrog/bamboo/configuration/ArtifactoryCollectBuildIssuesConfiguration.java [hunk 9]
```diff
@@ -40,6 +41,8 @@ public void populateContextForEdit(@NotNull Map<String, Object> context, @NotNul
         populateContextWithConfiguration(context, taskDefinition, FIELDS_TO_COPY);
         context.put("selectedServerId", context.get(CollectBuildIssuesContext.SERVER_ID_PARAM));
         context.put("serverConfigManager", serverConfigManager);
+        // Add default values to an existing task configuration.
+        AbstractArtifactoryConfiguration.populateDefaultBuildNameNumberInBuildContext(context);
```
[the remark this answers is not in the document]
Comment (barbelity) [reply src/main/java/org/jfrog/bamboo/configuration/ArtifactoryCollectBuildIssuesConfiguration.java:45]: Moved.

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

File: src/main/java/org/jfrog/bamboo/builder/BuildInfoHelper.java [hunk 10]
```diff
@@ -46,11 +45,15 @@
     private final String vcsRevision;
     private final String vcsUrl;
     private ServerConfig serverConfig;
+    private String buildName;
+    private String buildNumber;
 
-    private BuildInfoHelper(Map<String, String> env, String vcsRevision, String vcsUrl) {
+    private BuildInfoHelper(Map<String, String> env, String vcsRevision, String vcsUrl, String buildName, String buildNumber) {
```
[the remark this answers is not in the document]
Comment (barbelity) [reply src/main/java/org/jfrog/bamboo/builder/BuildInfoHelper.java]: Fixed.

**Pull request merged**